test/ng/directive/ngClassSpec.js

```diff
@@ -462,6 +462,47 @@ fdescribe('ngClass', function() {
     expect(e2.hasClass('odd')).toBeFalsy();
   }));
 
+
+  it('should keep track of old classes even if odd/even does not match `$index` atm',
```

Comment:
It might be because I was trying to solve it differently at first. Here is what was happening before this change:

1. Assume `ng-class-odd="foo"`, `$index = 0`, `foo = 'foo'`. The element has `className === 'foo'`.
2. If the `$index` and `foo` change simultaneously, say `$index = 1`, `foo = 'bar'`, then:
    1. The `attr[name]` watch kicks in first and it checks if `scope.$index & 1 === selector` (i.e. it uses the current value of `$index` to determine if it needs to update the classes). It decides it doesn't need to do anything, so at this point the element still has `className === 'foo'`.
    2. The `$index` watch kicks in. Since it determines that `$index & 1 !== old$index & 1`, it decides to remove the classes. But it tries to `scope.$eval(attr[name])` which now evaluates to `'bar'`. So the element will still have `className === 'foo'`.

Basically, the problem is that both watch-actions check against the current value of each other to determine what to do, which fails when both change in the same digest.
I.e. the `attr[name]` watch uses `scope.$index` (while it should use the previous index) and the `$index` watch uses `scope.$eval(attr[name])` (while it should use the previous value - the "old classes"). 

--
The commit fxes it by:

1. Moving the `$index` watch _before_ the `attr[name]` watch.
2. Using `oldVal` from the `$index` watch to add/remove.

This way, when the `$index & 1` changes, the `$index` watch will correctly add/remove the classes and then the `attr[name]` watch will update the classes (i.e. add the new ones) if necessary.

--
I hope this makes sense :smiley:

How about:

> It should add/remove the correct classes when the expression and `$index` change simultaneously